### PR TITLE
feat(stage): remove stage files

### LINKED ISSUE #4491: stage: remove stage files
Removes files from either an external (external cloud storage) or internal stage.
Synatx:
```
REMOVE stage [ PATTERN = '<regex_pattern>' ]
```

Where
```
stage ::=
    @[<stage_name>[/<file>]
```

### PR DESCRIPTION
**Pull request by Kikkon**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Support remove stage files

## Changelog

- New Feature

## Related Issues

Close #4491

## Test Plan
Unit Tests
Stateless Tests

### CODE REVIEW

File: tests/suites/1_stateful/00_copy/00_0000_copy_from_stage.sh [hunk 1]
```diff
@@ -66,7 +66,13 @@ done
 ## List stage use http API
 
 curl -s -u root: -XPOST "http://localhost:${QUERY_HTTP_HANDLER_PORT}/v1/query" --header 'Content-Type: application/json' -d '{"sql": "list @s1;"}'  | grep -o 'ontime_200.csv'
- 
+
+## Remove stage file
```
Comment (Xuanwo) [inline tests/suites/1_stateful/00_copy/00_0000_copy_from_stage.sh:70]: How about adding new test cases like `00_0000_remove_stage.sh`?

**Comment by Xuanwo** [comment]
Nice work!

File: tests/suites/1_stateful/00_copy/00_0000_copy_from_stage.sh (hunk 1 above)
Comment (Kikkon) [reply tests/suites/1_stateful/00_copy/00_0000_copy_from_stage.sh:70]: Sounds good. stateful test still has some bugs, I will fix them together.

File: query/src/sql/statements/statement_remove_user_stage.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+#[async_trait::async_trait]
+impl AnalyzableStatement for DfRemoveStage {
+    #[tracing::instrument(level = "info", skip(self, _ctx), fields(_ctx.id = _ctx.get_id().as_str()))]
+    async fn analyze(&self, _ctx: Arc<QueryContext>) -> Result<AnalyzedResult> {
+        let pattern = self.pattern.clone();
+        let path_vec: Vec<&str> = self.location.split('/').filter(|v| !v.is_empty()).collect();
+        //get file_name
+        let mut file_name = "".to_string();
+        if path_vec.len() > 1 {
```
Comment (sundy-li) [inline query/src/sql/statements/statement_remove_user_stage.rs:41]: The last path may be a directory ?

File: query/src/interpreters/interpreter_user_stage_remove.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        let plan = self.plan.clone();
+        let ctx = self.ctx.clone();
+
+        let mut files: Vec<String> = vec![];
+
+        if !plan.file_name.is_empty() {
+            files.push(plan.file_name);
+        }
+
+        if !plan.pattern.is_empty() {
```
Comment (sundy-li) [inline query/src/interpreters/interpreter_user_stage_remove.rs:60]: may be `else if ` is better

**Comment by sundy-li** [comment]
```
03_0000_remove_stage:                                                   [ FAIL ] - return code 126
, result:
/bin/sh: 1: /home/runner/work/databend/databend/tests/suites/1_stateful/03_remove/03_0000_remove_stage.sh: Permission denied

```

Need `chmod +x 03_0000_remove_stage.sh`.

File: tests/suites/1_stateful/00_copy/00_0000_copy_from_stage.sh [hunk 4]
```diff
@@ -72,3 +72,39 @@ curl -s -u root: -XPOST "http://localhost:${QUERY_HTTP_HANDLER_PORT}/v1/query" -
...
+echo "CREATE STAGE s1;" | $MYSQL_CLIENT_CONNECT
+## Remove internal stage file
+echo "remove @s1/ontime_200.csv.gz" | $MYSQL_CLIENT_CONNECT
+echo "list @s1 PATTERN = 'ontime.*'" | $MYSQL_CLIENT_CONNECT
+## Remove internal stage file with pattern
+echo "remove @s1  PATTERN = 'ontime.*'" | $MYSQL_CLIENT_CONNECT
+
+## Copy from named external stage
+echo "CREATE STAGE named_external_stage url = 's3://testbucket/admin/tempdata/' credentials=(aws_key_id='minioadmin' aws_secret_key='minioadmin');" | $MYSQL_CLIENT_CONNECT
+echo "list @named_external_stage PATTERN = 'ontime.*'" | $MYSQL_CLIENT_CONNECT
```
Comment (sundy-li) [inline tests/suites/1_stateful/00_copy/00_0000_copy_from_stage.sh:104]:
```suggestion
echo "list @named_external_stage PATTERN = 'ontime.*'" | $MYSQL_CLIENT_CONNECT | sort 
```

Comment (Kikkon) [reply tests/suites/1_stateful/00_copy/00_0000_copy_from_stage.sh:104]: get! thank you for help.

Comment (Kikkon) [reply tests/suites/1_stateful/00_copy/00_0000_copy_from_stage.sh:104]: But I did not find a similar `chmod +x 00_0000_copy_from_s3_location.sh ?

Comment (sundy-li) [reply tests/suites/1_stateful/00_copy/00_0000_copy_from_stage.sh:104]: It's to make `00_0000_copy_from_s3_location.sh` can be executed.

You need to execute it locally and commit to git.

See  https://linuxtect.com/what-is-chmod-x-command-in-linux/

**Review by Xuanwo — APPROVED** [review]
Nice work!

**Comment by BohuTANG** [comment]
This PR needs documentation, cc @soyeric128

**Pull request merged**